optaplanner-core/src/main/java/org/optaplanner/core/impl/score/buildin/hardsoft/HardSoftScoreInliner.java

```diff
@@ -16,62 +16,79 @@
 
 package org.optaplanner.core.impl.score.buildin.hardsoft;
 
-import java.util.function.Consumer;
+import java.util.Map;
 
-import org.optaplanner.core.api.score.Score;
 import org.optaplanner.core.api.score.buildin.hardsoft.HardSoftScore;
-import org.optaplanner.core.impl.score.inliner.IntWeightedScoreImpacter;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.impl.score.inliner.JustificationsSupplier;
 import org.optaplanner.core.impl.score.inliner.ScoreInliner;
+import org.optaplanner.core.impl.score.inliner.UndoScoreImpacter;
+import org.optaplanner.core.impl.score.inliner.WeightedScoreImpacter;
 
-public class HardSoftScoreInliner extends ScoreInliner<HardSoftScore> {
+public final class HardSoftScoreInliner extends ScoreInliner<HardSoftScore> {
 
-    protected int hardScore;
-    protected int softScore;
+    private int hardScore;
+    private int softScore;
 
-    protected HardSoftScoreInliner(boolean constraintMatchEnabled) {
-        super(constraintMatchEnabled);
+    protected HardSoftScoreInliner(Map<Constraint, HardSoftScore> constraintIdToWeightMap, boolean constraintMatchEnabled) {
+        super(constraintIdToWeightMap, constraintMatchEnabled, HardSoftScore.ZERO);
```

Comment:
This `Map<Constraint, HardSoftScore> constraintIdToWeightMap` is very similar to `ConstraintMatchTotal`

Reply:
It does something different . But I can remove it.

Instead, you'd have to do `ScoreInliner.buildScoreImpacter(Solution, Constraint)`, and the weight would be inferred at that point.
They are the same thing, they have the same effect, but maybe the second approach is better, as it is more local.

Reply:
Agreed that it's not CMT